[SPIR-V] Fix block sorting with irreducible CFG

Keenuts opened the pull request:
Block sorting was assuming reducible CFG. Meaning we always had a best node to continue with. Irreducible CFG makes breaks this assumption, so the algorithm looped indefinitely because no node was a valid candidate.

Fixes #116692

Keenuts committed:
[SPIR-V] Fix block sorting with irreducible CFG

Block sorting was assuming reducible CFG. Meaning we always
had a best node to continue with. Irreducible CFG makes breaks
this assumption, so the algorithm looped indefinitely because
no node was a valid candidate.

Fixes #116692

Signed-off-by: Nathan Gauër <brioche@google.com>

VyacheslavLevytskyy commented on llvm/lib/Target/SPIRV/SPIRVUtils.cpp:
```diff
@@ -525,6 +525,12 @@ size_t PartialOrderingVisitor::GetNodeRank(BasicBlock *BB) const {
+    // This block hasn't been ranked yet. Ignoring.
+    // This doesn't happen often, but when dealing with irreducible CFG, we have
+    // to rank nodes without knowing the rank of all their predecessors.
+    if (Iterator == BlockToOrder.end())
```
I don't understand how line 527 and 531 work together.

Keenuts replied:
They dont't, and I need better test coverage. Somehow I messed up and added this at the wrong place. But somehow all my tests now pass, which means either I lack coverage, or the code below is useless.

Keenuts replied:
Thanks for catching my mistake! That code is quite embarrassing to look at 😅

Keenuts replied:
All right, extended the test to also verify the rank value (used in the partial traversal function), and fixed the code. Thanks!

VyacheslavLevytskyy commented:
Does it make sense to add the reproducer from https://github.com/llvm/llvm-project/issues/116692, or those changes in the PR covers the reproducer's role in full?

Keenuts commented:
> Does it make sense to add the reproducer from #116692, or those changes in the PR covers the reproducer's role in full?

I added 2 versions of the reproducer in the tests:
 - 1 with the exact CFG (minus phi instructions, which doesn't matter for dominance)
 - 1 with the reduced CFG up to the bad pattern (to pin-point a specific pattern)

s-perron reviewed (commented):
I'm not sure if this works for irreducible regions because I am not sure what we are trying to accomplish.

This seems to be the strategy: if no node in `toVisit` has all of it predecessor ranked yet, then essentially randomly pick a node in `toVisit` as the next node.

Theoretically we could have:

A  -> B
A  -> C
A  -> D
B  -> C
B  -> D
C  -> B

In this case, you will get stuck with `toVisit = { B, C, D }`. Do we care if we choose `D` next, even if it is not part of the irreducible region? This would mean that `D` will have the same rank as `B` and `C`, and will have the earlier traversal number for the tie-breaker.

I tryed to find the restrictions on block order in spir-v. All I found was:

> The order of blocks in a function must satisfy the rule that blocks appear before all blocks they dominate.

If this is the only restriction, then we can do a much simpler algorithm: any pre-order traversal of the dominator tree.

I thought there was an restriction where a constructs has to be properly nested in each other for structured control flow. So we could not have H1, H2, M1, M2 where H<i> is a header with M<i> as its merge block. However, I cannot find that.

Keenuts commented:
> I'm not sure if this works for irreducible regions because I am not sure what we are trying to accomplish.

For irreducible sub-graphs, what we want is just to comply with the requirement that a block must be placed the ones it dominates in the binary.

> In this case, you will get stuck with `toVisit = { B, C, D }`. Do we care if we choose `D` next, even if it is not part of the irreducible region? 

In this case, no, we don't care. Maybe if we need a better fix, we could decide to isolate the irreducible region (here determine graph could be reducible if B-C were merged, hence pick should be between B and C).

> If this is the only restriction, then we can do a much simpler algorithm: any pre-order traversal of the dominator tree.

This would be fine for irreducible graph sorting. But the topological visitor used here is also used for structurizing.
What's important in such case is that loops belongs before the exit node.

Given this graph:
```
digraph {
   Entry -> Header;
   Header -> Continue;
   Header -> Exit;
   Continue -> Header;
}
```

The Dom-tree is as follows:
```
Entry -> Header
Header -> Exit
Header -> Continue
```
The header dominates both the exit and the continue block. But we MUST give C a lower rank than D so we can properly structurize the graph.
For this reason the algorithm is slightly tweaked to look for loops, and prioritize traversal of the loops.
(Algorithm is some kind of biased pre-order traversal of the D-tree)

s-perron commented:
> What's important in such case is that loops belongs before the exit node.

Why is this important? I thought that was part of the spec, but I cannot find it. Is it part of how the structurizer is implemented?

For changes to this PR, you should update the comments in SPIRVUtils.h to mention the case where it might fail to give a topological order, and why it does not matter. There should be an assert that when it fails we do not need to run the structurizer (we are not compiling for Vulkan).

Keenuts commented:
> > What's important in such case is that loops belongs before the exit node.
> 
> Why is this important? I thought that was part of the spec, but I cannot find it. Is it part of how the structurizer is implemented?

For the problem of BB sorting for binary output, I don't think there is a spec bit that says so.
It's really just a matter of how the structurizer works.

We have 3 parts:
 - the structurizer, which relies on the partial ordering visitor, and need the loop bit.
 - the partial ordering, which gives the node rank, and loop bit.
 - and the sorting, which only requires dominance ordering.

Because I had the partial ordering, which provided a super-set of the requirements for sorting, I used it to sort blocks.
But indeed we could split the two: have a sorting function which only relies on dominance pre-order, and the partial ordering, which is required for the structurizer.


> For changes to this PR, you should update the comments in SPIRVUtils.h to mention the case where it might fail to give a topological order

Sure, I can add a comment. But for irreducible, this order is not defined, so it does not "fail". It just gives 1 version of it.

> **EDIT:** Now that I think about it. I don't like that a class could silently fail to do what it is suppose to do. It could be confusing for people who try to us it in a different context.

Once again, IMO it doesn't fail. Topological sorting is not defined the graphs with cycles, even less for irreducible graphs. Hence this is a kinda-topological with a bias on the loop body ordering.


So I see 2 ways forward:
 - keep this as-is, using the kinda-topological-sorting to sort blocks.
 - reimplement the sorting function to only use pre-order so we stick to the spec requirements, and leave the visitor for the structurizer.

VyacheslavLevytskyy commented:
@Keenuts Let's please proceed with adding a comment explaining what are guarantees of the algorithm and what it tries to achieve, mentioning irreducible case. The PR fixes the issue https://github.com/llvm/llvm-project/issues/116692 so we should proceed but should not lose ideas/asserts from the comments.

Keenuts commented:
> @Keenuts Let's please proceed with adding a comment explaining what are guarantees of the algorithm and what it tries to achieve, mentioning irreducible case. The PR fixes the issue #116692 so we should proceed but should not lose ideas/asserts from the comments.

I'll send another PR which sorts the IR using a preorder traversal instead. This one will instead add tests to the visitor, and make irreducible case fail and not hang.
Just got side tracked on other tasks, but it's next on my list.

Keenuts committed:
change sorting to use reverse post order.

This commit also changes many tests as the new sorting has a slightly
different order:
 - second brcond operand goes first if possible.
 - switch operands are sorted last to first.

Keenuts committed:
restore test

Keenuts committed:
cleanup visitor changes

Signed-off-by: Nathan Gauër <brioche@google.com>

Keenuts committed:
clang-format

Signed-off-by: Nathan Gauër <brioche@google.com>

Keenuts committed:
revert minor changes

Signed-off-by: Nathan Gauër <brioche@google.com>

Keenuts commented:
Hello all!
This PR now leaves the visitor as-is, and only changes the `sortBlocks()` function to use a reverse port order iterator.
It also adds unit tests distinct from the ones for the visitor, testing block sorting only.

The PR to handle irreducible in the visitor is https://github.com/llvm/llvm-project/pull/117887

VyacheslavLevytskyy commented:
Irrelevant to this PR, but nevertheless -- this `CodeGen/SPIRV/debug-info/debug-type-basic.ll` recently started to fail in Github Actions and I'm not able to reproduce spirv-val complaint locally. Did you manage to reproduce that test case fail locally?

Keenuts commented:
> Irrelevant to this PR, but nevertheless -- this `CodeGen/SPIRV/debug-info/debug-type-basic.ll` recently started to fail in Github Actions and I'm not able to reproduce spirv-val complaint locally. Did you manage to reproduce that test case fail locally?

Same here, can see it's not related to this PR, but cannot reproduce it on my machine. Not sure why.
